### PR TITLE
Avoid unexpected concurrent access

### PR DESCRIPTION
**Pull request by linD026**
In chardev.c, chardev2.c and sleep.c files, the variable to
determine the exclusive access is an integer type, which leads
to the race condition.

In chardev.c and chardev2.c files, change the integer type of
"open_device_cnt" into atomic type.
In sleep.c, the integer type of "already_open" should also be atomic type.
And all the related operations are modified to atomic operations.

The reason why it is unnecessary to change the read operation is
`open_device_cnt` already prevents multiple access.

- sysprog21/lkmpg: https://github.com/sysprog21/lkmpg/issues/93

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
Change TeX document respectively. You might cite LWN articles: https://lwn.net/Articles/691128/

File: examples/chardev.c [hunk 1]
```diff
@@ -27,8 +27,9 @@ static ssize_t device_write(struct file *, const char __user *, size_t,
 /* Global variables are declared as static, so are global within the file. */
 
 static int major; /* major number assigned to our device driver */
-static int open_device_cnt = 0; /* Is device open?
-                                 * Used to prevent multiple access to device */
+static atomic_t open_device_cnt = ATOMIC_INIT(0); /* Is device open?
```
Comment (jserv) [inline examples/chardev.c:30]: Move comment before the declaration of `open_device_cnt`.

File: examples/.clang-format [hunk 2]
```diff
@@ -91,4 +92,5 @@ SpacesInParentheses: false
 SpacesInSquareBrackets: false
 Standard: Cpp03
 TabWidth: 4
-UseTab: Never
\ No newline at end of file
+UseTab: Never
```
Comment (jserv) [inline examples/.clang-format:95]: Don't change this line. Always minimize the necessary changes.

File: examples/example_atomic.c [hunk 3]
```diff
@@ -6,10 +6,10 @@
 #include <linux/module.h>
 
 #define BYTE_TO_BINARY_PATTERN "%c%c%c%c%c%c%c%c"
-#define BYTE_TO_BINARY(byte)                                                   \
-    (byte & 0x80 ? '1' : '0'), (byte & 0x40 ? '1' : '0'),                      \
-        (byte & 0x20 ? '1' : '0'), (byte & 0x10 ? '1' : '0'),                  \
-        (byte & 0x08 ? '1' : '0'), (byte & 0x04 ? '1' : '0'),                  \
+#define BYTE_TO_BINARY(byte)                                  \
```
Comment (jserv) [inline examples/example_atomic.c:9]: Why did you change the indention?

File: lkmpg.tex [hunk 4]
```diff
@@ -925,6 +925,17 @@ \subsection{chardev.c}
 Don't worry if you don't see what we do with the data we read into the buffer; we don't do much with it.
 We simply read in the data and print a message acknowledging that we received it.
 
+In the multiple-threaded case, without any protection, concurrent access to the same memory may cause the race condition, and will not preserve the performance.
```
Comment (jserv) [inline lkmpg.tex:928]: We usually say, "in the multi-threaded environment / use case"

Comment (jserv) [reply lkmpg.tex:928]: Change "cause" to "lead to"

See https://lexically.net/TimJohns/Kibbitzer/revis024.htm
* `lead to` is not used with a human subject, and does not appear in the passive.
* `lead to` is less direct than `cause`, implying a series of steps between cause and effect.

File: lkmpg.tex [hunk 5]
```diff
@@ -925,6 +925,17 @@ \subsection{chardev.c}
...
+In the kernel module, this problem also arises.
```
Comment (jserv) [inline lkmpg.tex:929]: Change to "this problem may happen due to ..."

File: lkmpg.tex [hunk 6]
```diff
@@ -925,6 +925,17 @@ \subsection{chardev.c}
...
+So, we have to let the instance exclusively access the module.
+When we open this file, we first check \cpp|open_device_cnt| is 0 or 1.
```
Comment (jserv) [inline lkmpg.tex:931]: Since the meaning of variable `open_device_cnt` has been changed, it would be better to rename the variable. Something like `is_device_open` can be more precise.

File: lkmpg.tex [hunk 7]
```diff
@@ -925,6 +925,17 @@ \subsection{chardev.c}
...
+If it is 1, deny the open operation.
+If \cpp|open_device_cnt| is 0, we atomically change to 1.
+Therefore, others cannot simultaneously open the same file with us.
+After that, we can successfully open the file.
+Here we use CAS (Compare And Swap) atomic operation to do read-modify-write.
```
Comment (jserv) [inline lkmpg.tex:936]: Improve the writing. What is your intention?

File: examples/example_atomic.c (hunk 3 above)
Comment (linD026) [reply examples/example_atomic.c:9]: Because in my ubuntu, when I generated the pdf file, the line break macro will go to the next line.



it can avoid by setting `AlignEscapedNewlines: Left`.

Comment (jserv) [reply examples/example_atomic.c:9]: Don't change the style in this pull request. You can submit another one to concentrate on further style tweaks.

File: lkmpg.tex [hunk 8]
```diff
@@ -925,6 +925,12 @@ \subsection{chardev.c}
 Don't worry if you don't see what we do with the data we read into the buffer; we don't do much with it.
 We simply read in the data and print a message acknowledging that we received it.
 
+In the multiple-threaded environment, without any protection, concurrent access to the same memory may lead to the race condition, and will not preserve the performance.
+In the kernel module, this problem may happen due to multiple instances access the resources.
```
Comment (jserv) [inline lkmpg.tex:929]: Change from "multiple instances access the resources" to
> multiple instances accessing the shared resources.

File: lkmpg.tex [hunk 9]
```diff
@@ -925,6 +925,12 @@ \subsection{chardev.c}
...
+So, we have to let the instance exclusively access the module.
```
Comment (jserv) [inline lkmpg.tex:930]: Change to
> Therefore, a solution is to enforce the exclusive access.

File: lkmpg.tex [hunk 10]
```diff
@@ -925,6 +925,12 @@ \subsection{chardev.c}
...
+Here we use CAS (Compare And Swap) atomic operation to check out the file is currently open by someone.
```
Comment (jserv) [inline lkmpg.tex:931]: Change to
> Atomic Compare-And-Swap (CAS) is used ...

File: lkmpg.tex [hunk 11]
```diff
@@ -925,6 +925,12 @@ \subsection{chardev.c}
...
+See more concurrency detail in the Avoiding Collisions and Deadlocks section.
```
Comment (jserv) [inline lkmpg.tex:932]: Use `\ref` for anchoring.

File: lkmpg.tex [hunk 12]
```diff
@@ -1158,6 +1164,8 @@ \section{Talking To Device Files}
 If you want to use ioctls in your own kernel modules, it is best to receive an official ioctl assignment, so if you accidentally get somebody else's ioctls, or if they get yours, you'll know something is wrong.
 For more information, consult the kernel source tree at \src{Documentation/userspace-api/ioctl/ioctl-number.rst}.
 
+Also, we need to be careful with the concurrent access that we mentioned in \verb|chardev.c| section.
```
Comment (jserv) [inline lkmpg.tex:1167]: This sentence is meaningless. We should emphasize on the risk and solutions.

File: lkmpg.tex [hunk 13]
```diff
@@ -1460,6 +1468,17 @@ \subsection{Atomic operations}
 
 \samplec{examples/example_atomic.c}
 
+Before the C11 standard added the built-in atomic types, the kernel already provided a small set of atomic types by using a bunch of tricky architecture-specific codes.
```
Comment (jserv) [inline lkmpg.tex:1471]: s/added/adopts/

File: lkmpg.tex [hunk 14]
```diff
@@ -1460,6 +1468,17 @@ \subsection{Atomic operations}
...
+Many discussions are talking about implementing atomic types of the kernel with C11 atomics.
```
Comment (jserv) [inline lkmpg.tex:1472]: This sentence is not really functional.

File: lkmpg.tex [hunk 15]
```diff
@@ -925,6 +925,13 @@ \subsection{chardev.c}
 Don't worry if you don't see what we do with the data we read into the buffer; we don't do much with it.
 We simply read in the data and print a message acknowledging that we received it.
 
+In the multiple-threaded environment, without any protection, concurrent access to the same memory may lead to the race condition, and will not preserve the performance.
+In the kernel module, this problem may happen due to multiple instances accessing the shared resources.
+Therefore, a solution is to enforce the exclusive access.
+Atomic Compare-And-Swap (CAS) is used, it check out the file is currently open by someone.
```
Comment (jserv) [inline lkmpg.tex:931]: Check the grammar.

File: lkmpg.tex [hunk 16]
```diff
@@ -925,6 +925,13 @@ \subsection{chardev.c}
...
+CAS will atomically compare the contents of the object with the expected value and, if they are the same, modifies the object into the desired value.
```
Comment (jserv) [inline lkmpg.tex:932]: Ditto. grammar!

File: lkmpg.tex [hunk 17]
```diff
@@ -925,6 +925,13 @@ \subsection{chardev.c}
...
+See more concurrency detail in the \ref{sec:synchronization} section.
```
Comment (jserv) [inline lkmpg.tex:933]: details

File: lkmpg.tex [hunk 18]
```diff
@@ -1158,6 +1165,9 @@ \section{Talking To Device Files}
 If you want to use ioctls in your own kernel modules, it is best to receive an official ioctl assignment, so if you accidentally get somebody else's ioctls, or if they get yours, you'll know something is wrong.
 For more information, consult the kernel source tree at \src{Documentation/userspace-api/ioctl/ioctl-number.rst}.
 
+Also, we need to be careful with the concurrent access to the shared resources that lead to the race condition.
+The solution is to enforce the exclusive access by using atomic Compare-And-Swap (CAS) that we mentioned at \ref{sec:org7ce767e} section.
```
Comment (jserv) [inline lkmpg.tex:1169]: Check the grammar.

File: lkmpg.tex [hunk 19]
```diff
@@ -925,6 +925,13 @@ \subsection{chardev.c}
 Don't worry if you don't see what we do with the data we read into the buffer; we don't do much with it.
 We simply read in the data and print a message acknowledging that we received it.
 
+In the multiple-threaded environment, without any protection, concurrent access to the same memory may lead to the race condition, and will not preserve the performance.
+In the kernel module, this problem may happen due to multiple instances accessing the shared resources.
+Therefore, a solution is to enforce the exclusive access.
+We use the Compare-And-Swap (CAS), the single atomic operation, to checks out whether the file is currently open by someone.
```
[the remark this answers is not in the document]
Comment (jserv) [reply lkmpg.tex:931]: Check the grammar.

File: lkmpg.tex [hunk 20]
```diff
@@ -925,6 +925,13 @@ \subsection{chardev.c}
 Don't worry if you don't see what we do with the data we read into the buffer; we don't do much with it.
 We simply read in the data and print a message acknowledging that we received it.
 
+In the multiple-threaded environment, without any protection, concurrent access to the same memory may lead to the race condition, and will not preserve the performance.
+In the kernel module, this problem may happen due to multiple instances accessing the shared resources.
+Therefore, a solution is to enforce the exclusive access.
+We use atomic Compare-And-Swap (CAS), the single atomic operation, to checks out whether the file is currently open by someone.
```
Comment (jserv) [inline lkmpg.tex:931]: Change "to checks out" to
> to determine

You shall boost your English grammar skills by regular writing.

File: lkmpg.tex [hunk 21]
```diff
@@ -925,6 +925,13 @@ \subsection{chardev.c}
...
+CAS compares the content of the target with the expected value and, if they are the same, modifies the content to the desired value.
```
Comment (jserv) [inline lkmpg.tex:932]: You can borrow the wording from Wikipedia: https://en.wikipedia.org/wiki/Compare-and-swap

Be aware of the statement "only if they are the same"

File: lkmpg.tex [hunk 22]
```diff
@@ -1158,6 +1165,9 @@ \section{Talking To Device Files}
 If you want to use ioctls in your own kernel modules, it is best to receive an official ioctl assignment, so if you accidentally get somebody else's ioctls, or if they get yours, you'll know something is wrong.
 For more information, consult the kernel source tree at \src{Documentation/userspace-api/ioctl/ioctl-number.rst}.
 
+Also, we need to be careful that concurrent access to the shared resources will lead to the race condition.
+The solution is using atomic Compare-And-Swap (CAS), which we mentioned at \ref{sec:org7ce767e} section, to enforce the exclusive access.
```
Comment (jserv) [inline lkmpg.tex:1169]: Change "sec:org7ce767e" to something mnemonic.

**Comment by jserv** [comment]
Thank @linD026 again for the great work. I amended the git commit messages for simplicity.
